Allow override CDN location

boschr opened the pull request:
Possibility to override CDN location if you want to host the assets somewhere else.

Currently we're not able to use `mui-tel-input` because of a (very) strict CSP header which blocks the requests to flagcdn.com.

jakeisnt commented:
I would like this change as well. I think `cdnLocation should never be optional`(will break URLs otherwise without a proper null check) and have a default argument of `'flagcdn.com'` in both places

viclafouch commented on src/components/Flag/Flag.tsx:
```diff
@@ -45,14 +51,14 @@ const Flag = ({ size = 'small', isoCode, countryName = '' }: FlagProps) => {
           <source
             type="image/png"
-            srcSet={`https://flagcdn.com/w${width}/${isoCodeFormatted}.png`}
+            srcSet={`${cdnLocation}/w${width}/${isoCodeFormatted}.png`}
```
I think it would be better to customize the all url. 

Something like : 
```tsx
<MuiTelInput flagUrls={({isoCode, size}) => {
   return {
      png: 'yourUrl' // can be a local file,
     webp: '....',
     jpg: '....'
   }
}} />
```

And you call `flagUrls` function for each Flag.

WDYT ?

boschr replied:
I like this suggestion! I'm going to implement this :)

boschr commented:
> I would like this change as well. I think `cdnLocation` should never be optional (will break URLs otherwise without a proper null check) and have a default argument of `'flagcdn.com'` in both places

`null` is not an accepted value from the type definition, `cdnLocation` can be `string | undefined`, but not `null`. But still, if you want to break the flags, you can use an unexisting location or just a random string like "abc".

I don't get why you would define a default value on each location (like now is done for the `flagSize`). To me there is no benefit in that.

jakeisnt commented:
> > I would like this change as well. I think `cdnLocation` should never be optional (will break URLs otherwise without a proper null check) and have a default argument of `'flagcdn.com'` in both places
> 
> `null` is not an accepted value from the type definition, `cdnLocation` can be `string | undefined`, but not `null`. But still, if you want to break the flags, you can use an unexisting location or just a random string like "abc".
> 
> I don't get why you would define a default value on each location (like now is done for the `flagSize`). To me there is no benefit in that.

Semantics of `null | undefined` aside, the type definitions allow for passing the value `undefined`, which is incorrect

boschr commented:
> > > I would like this change as well. I think `cdnLocation` should never be optional (will break URLs otherwise without a proper null check) and have a default argument of `'flagcdn.com'` in both places
> > 
> > 
> > `null` is not an accepted value from the type definition, `cdnLocation` can be `string | undefined`, but not `null`. But still, if you want to break the flags, you can use an unexisting location or just a random string like "abc".
> > I don't get why you would define a default value on each location (like now is done for the `flagSize`). To me there is no benefit in that.
> 
> Semantics of `null | undefined` aside, the type definitions allow for passing the value `undefined`, which is incorrect

That is correct, you can fill in `undefined` as a value, but in the first place why would you do that? And second, it doesn't do anything, it still falls back to the default value. To me, this is exactly how optional props should work.

viclafouch commented on src/components/Flag/Flag.tsx:
```diff
@@ -27,11 +28,24 @@ const getSourceByIsoCode = (isoCode: MuiTelInputCountry | null) => {
   const sourceFound = getSourceByIsoCode(isoCode)
   // see https://flagpedia.net/download/api for the valid width
   const width = size === 'small' ? 40 : 80
+  const { webp, png } = getFlagSources(isoCodeFormatted, width)
```
Nah, just map on keys of the object, you could have more format (jpg, jpeg, etc...)

boschr replied:
I can do that, but then we need to extend the return of `getFlagSources` with the source type right?

viclafouch replied:
LGTM

viclafouch commented on src/index.types.ts:
```diff
@@ -15,6 +15,14 @@ export type FlagSize = `small` | 'medium'
+  isoCode: string,
+  width: 40 | 80
+) => {
+  png: string
```
maybe a cdn provides only jpg file for example. You should manage all cases.

viclafouch commented on src/index.types.ts:
```diff
@@ -15,6 +15,14 @@ export type FlagSize = `small` | 'medium'
 export type MuiTelInputReason = 'country' | 'input'
 
+export type GetFlagSources = (
+  isoCode: string,
```
I don't think it's "just" a string

viclafouch commented on src/components/Flag/Flag.tsx:
```diff
@@ -27,11 +28,24 @@ const getSourceByIsoCode = (isoCode: MuiTelInputCountry | null) => {
+}: FlagProps) => {
   const isoCodeFormatted = isoCode ? isoCode.toLowerCase() : ''
   const sourceFound = getSourceByIsoCode(isoCode)
   // see https://flagpedia.net/download/api for the valid width
```
you can remove this

viclafouch commented on src/components/Flag/Flag.tsx:
```diff
@@ -27,11 +28,24 @@ const getSourceByIsoCode = (isoCode: MuiTelInputCountry | null) => {
   const isoCodeFormatted = isoCode ? isoCode.toLowerCase() : ''
   const sourceFound = getSourceByIsoCode(isoCode)
   // see https://flagpedia.net/download/api for the valid width
   const width = size === 'small' ? 40 : 80
```
nah, 40 or 80 is not valid anymore as there is no only 1 cdn.. You should let the developer choose the `width` himself.

viclafouch commented on src/components/Flag/Flag.tsx:
```diff
@@ -27,11 +28,24 @@ const getSourceByIsoCode = (isoCode: MuiTelInputCountry | null) => {
+  size = 'small',
+  isoCode,
+  countryName = '',
+  getFlagSources = defaultGetFlagSources
```
I'm not sure, but I would say you break the `memo` wrapper as the `getFlagSources` get a new reference on every render

boschr replied:
I'm also not 100% sure, but i don't think this is the case. As how i see this: `defaultGetFlagSources` is defined outside the component, so it will only be defined once on build time and will never get a new reference.

viclafouch replied:
Yes but `getFlagSources` if it's defined, will get a new reference on every render in user app

boschr replied:
Hmm, ok, this is out of my knowledge.. Can you suggest a solution for this?

viclafouch replied:
I think we can let this as it is, and provide more information in the documentation

boschr replied:
Ok, then i will leave it as is. Provide more informatie? Updating the API Reference is that enough? (that what I just did). Let me know if you miss something.

boschr commented:
I changed the result of the `getFlagSources` to an array instead of an object, but you will see.

Further I implemented all your feedback as good as possible.

If there is a source defined with a `src` attribute it will fallback to a regular image. This is needed for inline SVG's (as the current implementation is doing for AC and TA) which are not meant to be in a `<picture>` as i understand from: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/source#srcset

viclafouch commented on src/components/Flag/Flag.tsx:
```diff
@@ -2,63 +2,81 @@ import React from 'react'
   return (
     <Styled.Flag data-testid={isoCode} className="MuiTelInput-Flag">
-      {sourceFound ? (
+      {regularImage
```
Run a little script `npm run lint:fix`, prettier is not working here I think :D

viclafouch commented on src/index.types.ts:
```diff
@@ -15,6 +15,22 @@ export type FlagSize = `small` | 'medium'
 
 export type MuiTelInputReason = 'country' | 'input'
 
+type GetFlagSourcesOutput = {
```
FlagSource, not `get`, is not a function or smth

viclafouch commented on src/components/Flag/Flag.tsx:
```diff
@@ -2,59 +2,87 @@ import React from 'react'
 import unknownFlag from '@assets/unknown-flag.png'
```
Could you please create an example in a new storybook component plz ?

boschr replied:
I updated the API Reference section in the docs. Is that what you mean?

boschr replied:
Hmm, i'm lost, I have no idea what to add and where. This is the storybook file you are referencing?
`src/index.stories.tsx`?

But what is a "Template"? Is it a `StoryFn` like "Primary"? What do you want me to put there? I also don't get the idea of that Primary page to be honest.

viclafouch replied:
Ok no worries, I'm taking care of it

viclafouch replied:
What other CDN you have for example plz?

boschr replied:
Within our company I already created a CDN before for hosting other assets. Im going to use this CDN also for the flags, but it's meant to be to communicated or used for the outside world unfortunately. I assume you want to use an other CDN as an example in the storybook?

About the `flagSize`, as you know I didn't touch that part, but i agree that it's not very UI friendly. Could be another improvement :) but if people are using it, it would be a breaking change :(

boschr replied:
I agree that is would be (a bit) more easy for devs, but also it's not that hard at the moment right?

Not sure what the pro's and con's are with the `<picture />` tag. From what I read (w3schools) is that the picture tag especially useful is for responsive designs. And that is not the case for the flags, so yes could be refactored, but don't have a strong opinion.
